### PR TITLE
Add rewrite base to generated .htaccess rules

### LINKED ISSUE #40696: [QA] hardened .htaccess breaks index.php-less setup when ownCloud URL has a subfolder
Seen with 10.12.0 (regression)

* Setup a fresh owncloud server running in a folder, e.g. https://localhost/breakme
* Login as e.g. admin user at the Web UI. Folder listing with skeleton files appears. OK
* Connect a desktop client. Initial sync succeeds. OK
* Change the setup to index.php - less URLs like this:
   ```
   occ config:system:set htaccess.RewriteBase --value /breakme
   occ maintenance:update:htaccess
   ```

* Login at the web UI no longer shows files. BAD
* Desktop client fails to sync. BAD
  `23-03-24 00:35:13:402 [ info sync.httplogger ]:	"88320329-5295-4952-8197-cd963ca3d0af: Response: PROPFIND 405 (Error: Error transferring https://localhost/breakme/remote.php/dav/files/admin/ - server replied: Method Not Allowed,177ms)`

Only the combination of both "index.php-less setup" and "base URL with a folder" is affected. Either one alone, or none of the two works fine.

### PR DESCRIPTION
**Pull request by IljaN**
## Description
If htaccess.RewriteBase is set we need to prepend it in the generated rules.

## Related Issue

- Fixes https://github.com/owncloud/core/issues/40696
- Related https://github.com/owncloud/core/pull/40584

## Motivation and Context
The [hardened .htaccess fix](https://github.com/owncloud/core/pull/40584) caused a regression for installations which have owncloud in a subfolder.

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

File: lib/private/Setup.php [hunk 1]
```diff
@@ -491,26 +491,26 @@ public static function updateHtaccess() {
...
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBase/core/ajax/update\\.php";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBase/status\\.php$";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBase/ocs/v1\\.php";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBase/ocs/v2\\.php";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBase/updater/";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBase/ocs-provider/";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBase/ocm-provider/";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBase/\\.well-known/(acme-challenge|pki-validation)/.*";
 			$content .= "\n  RewriteRule . index.php [PT,E=PATH_INFO:$1]";
 			$content .= "\n  RewriteBase " . $rewriteBase;
```
Comment (IljaN) [inline lib/private/Setup.php:515]: Hmhh this would probably dont put a "/" here

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiSharePublicLink2-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/38112/60

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiShareManagementBasicToShares-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/38112/54

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiShareManagementToRoot-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/38112/51

**Review by jnweiger — CHANGES_REQUESTED** [review]
A literal unescaped $rewriteBase can destroy the regexp. 
We need to apply proper regexp escaping, this is a security relevant.

**Comment by IljaN** [comment]
We could do this in a addtional PR by sanitizing $rewriteBase with [preg_quote](https://www.php.net/manual/en/function.preg-quote.php). But IMO offtopic for this PR. A path shouldn't contain any Regex characters and is coming from the trusted config anyways, no?

**Comment by IljaN** [comment]
Fail: https://drone.owncloud.com/owncloud/core/38114/172/15

**Comment by jnweiger** [comment]
> We could do this in a addtional PR by sanitizing $rewriteBase with [preg_quote](https://www.php.net/manual/en/function.preg-quote.php). But IMO offtopic for this PR. A path shouldn't contain any Regex characters and is coming from the trusted config anyways, no?

No, I don't think so. This path is perfectly legal, imho.
occ config:system:set htaccess.RewriteBase --value '/funny$nameing-scheme-*-but+we+like+it'


Pfft. On another thought, we don't quote strings in .htaccess, that means it breaks apart, when RewriteBase contains whitespace...

File: lib/private/Setup.php [hunk 2]
```diff
@@ -492,25 +492,31 @@ public static function updateHtaccess() {
 		// Add rewrite rules if the RewriteBase is configured
 		$rewriteBase = $config->getSystemValue('htaccess.RewriteBase', '');
 		if ($rewriteBase !== '') {
+			$rewriteBase_re = \preg_quote($rewriteBase_re);
```
Comment (jnweiger) [inline lib/private/Setup.php:495]:
```suggestion
			$rewriteBase_re = \preg_quote($rewriteBase);
```

Comment (IljaN) [reply lib/private/Setup.php]: Okay definetly makes sense for "-" characters. Thx

**Comment by ho4ho** [comment]
Fix #40686 also

**Review by jnweiger — APPROVED** [review]

File: lib/private/Setup.php [hunk 3]
```diff
@@ -492,25 +492,30 @@ public static function updateHtaccess() {
...
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/public\\.php";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/cron\\.php";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/core/ajax/update\\.php";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/status\\.php$";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/ocs/v1\\.php";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/ocs/v2\\.php";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/updater/";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/ocs-provider/";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/ocm-provider/";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/\\.well-known/(acme-challenge|pki-validation)/.*";
```
Comment (ho4ho) [inline lib/private/Setup.php:518]:
```suggestion
			$content .= "\n  RewriteCond %{REQUEST_URI} !^/\\.well-known/(acme-challenge|pki-validation)/.*";
```

Thinking always needing at root and not by subfolder.

**Review by jnweiger — APPROVED** [review]

**Pull request merged**